x/reports/keeper/msg_server.go

```diff
@@ -0,0 +1,284 @@
+package keeper
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+
+	"github.com/desmos-labs/desmos/v3/x/reports/types"
+	subspacestypes "github.com/desmos-labs/desmos/v3/x/subspaces/types"
+)
+
+type msgServer struct {
+	Keeper
+}
+
+// NewMsgServerImpl returns an implementation of the stored MsgServer interface for the provided keeper
+func NewMsgServerImpl(keeper Keeper) types.MsgServer {
+	return &msgServer{Keeper: keeper}
+}
+
+var _ types.MsgServer = &msgServer{}
+
+// CreateReport defines the rpc method for Msg/CreateReport
+func (k msgServer) CreateReport(goCtx context.Context, msg *types.MsgCreateReport) (*types.MsgCreateReportResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+
+	// Check if the subspace exists
+	if !k.HasSubspace(ctx, msg.SubspaceID) {
+		return nil, sdkerrors.Wrapf(sdkerrors.ErrInvalidRequest, "subspace with id %d not found", msg.SubspaceID)
+	}
+
+	// Check if the reason exists
+	if !k.HasReason(ctx, msg.SubspaceID, msg.ReasonID) {
+		return nil, sdkerrors.Wrapf(sdkerrors.ErrInvalidRequest, "reason with id %d not found inside subspace %d", msg.ReasonID, msg.SubspaceID)
+	}
+
+	// Check the permission to report
+	if !k.HasPermission(ctx, msg.SubspaceID, msg.Reporter, subspacestypes.PermissionReportContent) {
+		return nil, sdkerrors.Wrap(subspacestypes.ErrPermissionDenied, "you cannot report content inside this subspace")
+	}
+
+	// Get the next report id
+	reportID, err := k.GetNextReportID(ctx, msg.SubspaceID)
+	if err != nil {
+		return nil, err
+	}
+
+	// Create and validate the report
+	report := types.NewReport(
+		msg.SubspaceID,
+		reportID,
+		msg.ReasonID,
+		msg.Message,
+		msg.Target.GetCachedValue().(types.ReportTarget),
+		msg.Reporter,
+		ctx.BlockTime(),
+	)
+	err = k.ValidateReport(ctx, report)
+	if err != nil {
+		return nil, err
+	}
+
+	// Store the report
+	k.SaveReport(ctx, report)
+
+	// Update the id for the next report
+	k.SetNextReportID(ctx, msg.SubspaceID, report.ID+1)
+
+	ctx.EventManager().EmitEvents(sdk.Events{
+		sdk.NewEvent(
+			sdk.EventTypeMessage,
+			sdk.NewAttribute(sdk.AttributeKeyModule, types.AttributeValueCategory),
+			sdk.NewAttribute(sdk.AttributeKeyAction, sdk.MsgTypeURL(msg)),
+			sdk.NewAttribute(sdk.AttributeKeySender, msg.Reporter),
+		),
+		sdk.NewEvent(
+			types.EventTypeCreateReport,
+			sdk.NewAttribute(types.AttributeKeySubspaceID, fmt.Sprintf("%d", msg.SubspaceID)),
+			sdk.NewAttribute(types.AttributeKeyReportID, fmt.Sprintf("%d", report.ID)),
+			sdk.NewAttribute(types.AttributeKeyReporter, msg.Reporter),
```

Comment:
Should we add target to the event as well?

Reply:
It would be hard to do so since the target is an interface and would be serialized as a JSON inside a string. Not the best thing in my opinion

Reply:
I don't think it's necessary. The target of a report can easily be retrieved by taking the subspace and report ids (both returned inside the event) and using the `Report` query to get all the details.

Reply:
In order to solve this problem, as well as to implement the query described [here](https://github.com/desmos-labs/desmos/pull/860#discussion_r886616253), I've added a new `reportedEvent` that is emitted when a report is created. 

If a post is reported, the event will contain the id of the post along with the subspace id and the reporter address.

If a user is reported, the event will instead contain the address of the reported user, the subspace id and the reporter address. 

This should allow to: 

1. query all the reports that were made to a now deleted post; 
2. query all the reports to a post/user that have been made from a specific user.